### PR TITLE
Improve Mark formatting

### PR DESCRIPTION
**Pull request by geriux**
**Related PRs:**
- https://github.com/wordpress-mobile/gutenberg-mobile/pull/6517
- https://github.com/WordPress/gutenberg/pull/57650
- https://github.com/wordpress-mobile/WordPress-iOS/pull/22425

This PR fixes several issues related to the Mark formatting.

- It completes the missing functionality implementation to fully support the Mark tag.
- Adds `MarkStringAttributeConverter` which will handle toggling the format as well as returning the expected HTML format to Gutenberg, by removing the extra spaces. This was causing issues like history changes or showing the active style indicator when it shouldn't.
- Updates `MarkFormatter` to set the expected CSS attributes as well as adding a `color` param.  
- Adds `preprocessMarkForInsertion` that will handle cases like auto-corrected strings, it will check if the new auto-corrected words need to have the Mark representation along with `adjustAttributesForMark`.
- It also fixes `toggleMark` by allowing passing a `color` param and to be able to change colors while editing a Mark formatting or resetting the existing format.

## To test

Check this [PR description](https://github.com/WordPress/gutenberg/pull/57650) under the `Test` section. Where there are builds linked as well.

### CODE REVIEW

**Comment by geriux** [comment]
Hey @twstokes 👋 I was wondering if you could help me out with a review for this PR since part of it uses the same code you added for the Heading fix for the highlighting text feature.

No worries if you can't, let me know either way. Thanks! 🙇

**Comment by twstokes** [comment]
> Hey @twstokes 👋 I was wondering if you could help me out with a review for this PR since part of it uses the same code you added for the Heading fix for the highlighting text feature.
> 
> No worries if you can't, let me know either way. Thanks! 🙇

Hey @geriux! 👋  I'll be glad to. 👍

**Comment by geriux** [comment]
> > Hey @twstokes 👋 I was wondering if you could help me out with a review for this PR since part of it uses the same code you added for the Heading fix for the highlighting text feature.
> > No worries if you can't, let me know either way. Thanks! 🙇
> 
> Hey @geriux! 👋 I'll be glad to. 👍

🎉 thank you!

File: Aztec/Classes/TextKit/TextStorage.swift [hunk 1]
```diff
@@ -253,6 +254,46 @@ open class TextStorage: NSTextStorage {
...
     }
 
+    /// Preprocesses an attributed string that is missing a `markHtmlRepresentation` attribute for insertion in the storage, this reuses the same behavior as preprocessHeadingsForInsertion
+    ///
+    /// - Important: This method adds the `markHtmlRepresentation` attribute if it determines the string should contain it.
+    ///  This works around a problem where autocorrected text didn't contain the attribute. This may change in future versions.
+    ///
+    /// - Parameters:
+    ///     - attributedString: the string we need to preprocess.
+    ///
```
Comment (twstokes) [inline Aztec/Classes/TextKit/TextStorage.swift:269]: Nit: Missing param for `_ range: NSRange`.

File: Aztec/Classes/TextKit/TextStorage.swift [hunk 2]
```diff
@@ -253,6 +254,46 @@ open class TextStorage: NSTextStorage {
...
+    /// - Returns: the preprocessed string.
+    ///
+    fileprivate func preprocessMarkForInsertion(_ attributedString: NSAttributedString, _ range: NSRange) -> NSAttributedString {
+        guard textStore.length > 0, attributedString.length > 0, range.endLocation != range.lowerBound else {
```
Comment (twstokes) [inline Aztec/Classes/TextKit/TextStorage.swift]: Is `range.endLocation != range.lowerBound` equivalent to `range.length > 0`?

`range.endLocation` is the same as `range.location` + `range.length`, and I believe `range.lowerBound` is equivalent to `range.location`. So it seems they can only be equivalent if length is zero, but I may be missing something!

File: Aztec/Classes/TextKit/TextStorage.swift [hunk 3]
```diff
@@ -253,6 +254,46 @@ open class TextStorage: NSTextStorage {
...
+            return attributedString
+        }
+        var startAt = 0
+
+        if range.endLocation > range.lowerBound {
```
Comment (twstokes) [inline Aztec/Classes/TextKit/TextStorage.swift]: I may need some help on this one as well - is there an instance where the end location won't be higher than the lower bound for lengths > 0?

File: Aztec/Classes/Formatters/Implementations/MarkFormatter.swift [hunk 4]
```diff
@@ -24,8 +24,8 @@ class MarkFormatter: AttributeFormatter {
     func remove(from attributes: [NSAttributedString.Key: Any]) -> [NSAttributedString.Key: Any] {
         var resultingAttributes = attributes
 
+        resultingAttributes[.foregroundColor] = placeholderAttributes![.foregroundColor]
```
Comment (twstokes) [inline Aztec/Classes/Formatters/Implementations/MarkFormatter.swift]: Are we sure `placeholderAttributes` is safe to force unwrap every time or should we do the `guard let` pattern?

**Review by twstokes — COMMENTED** [review]
👋 @geriux! I had a few questions in the PR, but otherwise the tests from the [Gutenberg PR](https://github.com/WordPress/gutenberg/pull/39454) passed. 👍 

I did notice a few things that looked like bugs that I wanted to share. They may not be due to this PR and you may already be aware of them. Unfortunately, only two were reproducible.

1. When viewing the HTML output, I found that one of my `<mark>` elements had two `color` properties. [couldn't reproduce]

Two color properties

2. The formatting bar retained the highlighted color even when it wasn't being applied. [reproducible]

https://user-images.githubusercontent.com/2092798/159201417-323294a1-d06b-471a-9752-92168eff178d.MP4

3. Starting with an empty paragraph block, choosing a highlight color, and typing resulted in no highlighted text. It was as if I'd never chosen the color. I didn't get a chance to grab a video of this one. [couldn't reproduce]

4. Text color didn't match the custom color chosen. I confirmed this also happens in the latest TestFlight version 19.4.0.3. This was a simple site running the Blank Canvas theme. It only happened with custom colors chosen at the bottom of the picker, not the custom color picker in the "Theme" section. [reproducible]

https://user-images.githubusercontent.com/2092798/159201625-ea512f9b-2453-4242-9a63-677d5fed831a.MP4

**Comment by twstokes** [comment]
👋 Hey @geriux! Is there anything I can do to help with this PR or have we decided not to pursue it for now? Thanks!

**Comment by geriux** [comment]
> 👋 Hey @geriux! Is there anything I can do to help with this PR or have we decided not to pursue it for now? Thanks!

Hey @twstokes Thank you for reaching out! Due to other priorities, I couldn't focus on addressing your feedback but I will go back to this so we can include this fix. 🙇

**Comment by twstokes** [comment]
> > 👋 Hey @geriux! Is there anything I can do to help with this PR or have we decided not to pursue it for now? Thanks!
> 
> Hey @twstokes Thank you for reaching out! Due to other priorities, I couldn't focus on addressing your feedback but I will go back to this so we can include this fix. 🙇

Sounds good!

**Comment by geriux** [comment]
Hey there @twstokes 👋 I'm sorry this took so long to move forward but I've introduced new changes that should fix the issues you mentioned above 🚀

File: Aztec/Classes/Formatters/Implementations/MarkFormatter.swift [hunk 5]
```diff
@@ -4,28 +4,43 @@ import UIKit
...
 
-        var representationToUse = HTMLRepresentation(for: .element(HTMLElementRepresentation.init(name: "mark", attributes: [])))
+        // Setting the HTML representation
+        var representationToUse = HTMLRepresentation(for: .element(HTMLElementRepresentation.init(name: "mark", attributes: [styleAttribute, classAttribute])))
         if let requestedRepresentation = representation {
             representationToUse = requestedRepresentation
         }
         resultingAttributes[.markHtmlRepresentation] = representationToUse
 
+        if textColor != nil {
```
Comment (twstokes) [inline Aztec/Classes/Formatters/Implementations/MarkFormatter.swift]: Optional nit: I'd consider an `if let` here to unwrap `textColor` safely so we don't have to force unwrap `textColor` later. It should be safe as is, just a preferred idiom.

File: Aztec/Classes/Formatters/Implementations/MarkFormatter.swift [hunk 6]
```diff
@@ -4,28 +4,43 @@ import UIKit
 class MarkFormatter: AttributeFormatter {
 
     var placeholderAttributes: [NSAttributedString.Key: Any]?
+    var textColor: String?
+    var defaulTextColor: UIColor?
```
Comment (twstokes) [inline Aztec/Classes/Formatters/Implementations/MarkFormatter.swift]: Optional nit:
```suggestion
    var defaultTextColor: UIColor?
```

**Review by twstokes — APPROVED** [review]
Great work @geriux! 🚀 This makes sense to me, and I added two comments that are optional nits. 

I ran through a couple tests, but mostly relied on David and Carlos' results [in this PR](https://github.com/WordPress/gutenberg/pull/57650). If you think I should do all of the manual tests as well I'd be happy to.

File: Aztec/Classes/Formatters/Implementations/MarkFormatter.swift (hunk 5 above)
Comment (geriux) [reply Aztec/Classes/Formatters/Implementations/MarkFormatter.swift]: Thanks for the suggestion! Addressed in https://github.com/wordpress-mobile/AztecEditor-iOS/pull/1352/commits/28e62ab8f254eec55b87e4a5607e2c9d4a9887d4

File: Aztec/Classes/Formatters/Implementations/MarkFormatter.swift (hunk 6 above)
Comment (geriux) [reply Aztec/Classes/Formatters/Implementations/MarkFormatter.swift]: Nice catch! Updated in https://github.com/wordpress-mobile/AztecEditor-iOS/pull/1352/commits/7b9cc634d514835a0fd92d7743123211b06da4ee

**Review by twstokes — APPROVED** [review]
Looks great! 🙇

**Pull request merged**